executable-spec/src/Cardano/Ledger/Spec/STS/Update/Tallysip.hs

```diff
@@ -0,0 +1,222 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Ledger.Spec.STS.Update.Tallysip where
+
+import           Data.Map.Strict (Map, (!))
+import qualified Data.Map.Strict as Map
+import           Data.Set as Set (Set)
+import qualified Data.Set as Set
+import           GHC.Generics (Generic)
+import           Data.AbstractSize (HasTypeReps)
+
+import           Control.State.Transition (Embed, Environment, IRC (IRC),
+                     PredicateFailure, STS, Signal, State, TRC (TRC),
+                     initialRules, judgmentContext,
+                     trans, transitionRules, wrapFailed, (?!))
+import           Ledger.Core (Slot, addSlot, (⨃), (∈), (∉), dom)
+import qualified Ledger.Core as Core
+import           Cardano.Crypto.Hash (HashAlgorithm)
+
+import qualified Cardano.Ledger.Spec.STS.Update.Data as Data
+
+-- | STS for tallying the votes of a single SIP
+data TALLYSIP hashAlgo
+
+data Env hashAlgo
+ = Env { currentSlot :: !Slot
+       , sipdb :: !(Map (Data.SIPHash hashAlgo) (Data.SIP hashAlgo))
+       , ballots :: !(Map (Data.SIPHash hashAlgo) (Map Core.VKey Data.Confidence))
+       , vThreshold :: !Data.VThreshold
+       , stakeDist :: !(Map Core.VKey Data.Stake)
```

Comment:
This should be the hash of a key. As far as I understand cryptographers discouraged us from the use of keys in ordered containers. I don't know if you can change this in this PR or we need to overhaul the entire specs so that only key hashes are used in ordered containers.

Reply:
Yes you are right, we have discussed this again in the past.
In what sense is `Map` an ordered container?

I think that if we are going to fix this, then we should do it completely/globally and not partially. So I suggest we wait when the corresponding task in Zenhub takes priority.

Reply:
> In what sense is Map an ordered container?

Basically it relies on its keys being an instance of `Ord`.